articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -216,7 +216,27 @@ Furthermore, service code is upgraded one upgrade domain at a time. So, during a
 > While you can modify the schema of a key, you must ensure that your key's hash code and equals algorithms are stable. If you change how either of these algorithms operate, you will not be able to look up the key within the reliable dictionary ever again.
 > .NET Strings can be used as a key but use the string itself as the key--do not use the result of String.GetHashCode as the key.
 
-Alternatively, you can perform what is typically referred to as a two upgrade. With a two-phase upgrade, you upgrade your service from V1 to V2: V2 contains the code that knows how to deal with the new schema change but this code doesn't execute. When the V2 code reads V1 data, it operates on it and writes V1 data. Then, after the upgrade is complete across all upgrade domains, you can somehow signal to the running V2 instances that the upgrade is complete. (One way to signal this is to roll out a configuration upgrade; this is what makes this a two-phase upgrade.) Now, the V2 instances can read V1 data, convert it to V2 data, operate on it, and write it out as V2 data. When other instances read V2 data, they do not need to convert it, they just operate on it, and write out V2 data.
+Alternatively, you can perform a multi-phase upgrade. 
+1. Deploy a new service version that
+    - has both the original version, V1 of the data contracts, and the new version, V2 of the data contracts.
+    - During this phase, the service must continue accessing data in the original, V1 collection using V1 data contracts.
+2. Deploy a new service version where
+    - a new, V2 collection iscreated;
+    - all add, update and delete operations are performed on both V1 and V2 collections in a single transaction;

```

Comment:
In this phase, in order to start doing Add/update/delete on both, you'd need to already have all of V1 from a point in time copied to V2, right? 

Also how do we actually do the Copy from V1 to V2 in the first place? If you are doing it live, with an enumeration from a snapshot transaction, V2 at that point is up to date with the state as of the LSN at the start of the Txn. So there will be a gap of LSNs applied since the start of that copy operation which need to be caught up...

Reply:
Ok, so I kind of get it. From a fresh dictionary we can do an "AddOrUpdate" in V2 for both Adds and Updates in V1. E.g if V2 didn't have the key yet and it was Updated, you would do an Add instead.

So while we add/update/delete the keys, we are concurrently doing a full RC snapshot enumeration which is comparing the stores and adding/updating/deleting keys to match.

I wonder if we could utilize Reliable Dictionary change notifications to simplify this flow. It would be a lot less efficient than doing everything twice in every transaction, but it might benefit from having some generic plug and play sample solution which customers could easily use.
https://learn.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-notifications#reliable-dictionary-notifications

Reply:
Good points. I also realized that the copy process should avoid resurrecting keys that were deleted from V1 since the copy process started aren't resurrected in the V2 from an old snapshot. I've expanded the description. Does it make sense now?

Reply:
@olegsych did you take a look at potentially using Change notifications (linked above) to simplify the mirroring of write operations from V1 to V2? I noticed that in the issue you linked in the other commment: https://github.com/microsoft/service-fabric/issues/676#issuecomment-439177045 

That Zuhair also suggested something similar. He crossed it out because the copy wouldn't work in the Rebuild phase, but it's not clear to me that wiring up the write mirroring via Change notifications wouldn't work. If we got something like that figured out, it would greatly simplify the code changes for the user to make, and we could provide a code snippet which would be usable in a general case by most customers. I'm not too familiar with their implementation though, so I'm not sure if it would actually work.

Reply:
No, I have not. Are change notifications transactional?